framework/Shell/Actions/TAppAction.php

```diff
@@ -0,0 +1,57 @@
+<?php
+/**
+ * TAppAction class file
+ *
+ * @author Brad Anderson <belisoful@icloud.com>
+ * @link https://github.com/pradosoft/prado
+ * @license https://github.com/pradosoft/prado/blob/master/LICENSE
+ * @package Prado\Shell\Actions
+ */
+
+namespace Prado\Shell\Actions;
+
+use Prado\Shell\TShellInterpreter;
+use Prado\Shell\TShellAction;
+
+/**
+ * Creates and run Prado application specific commands.
+ *
+ * @author Brad Anderson <belisoful[at]icloud[dot]com>
+ * @package Prado\Shell\Actions
+ * @since 4.2.0
+ */
+class TAppAction extends TShellAction
+{
+	protected $action = 'app';
+	protected $parameters = [];
+	protected $optional = ['directory', 'app-action'];
+	protected $description = 'Initializes the Prado application in the given [directory] and performs the app action';
+
+	/**
+	 * @param array $args parameters
+	 * @return bool
+	 */
+	public function performAction($args)
+	{
+		$app = null;
+		if (count($args) == 1) {
+			$args[1] = '.';
+		}
+		if (false === ($xml = $this->getAppConfigFile($args[1]))) {
+			echo "** Application '$args[1]' was not found. \n";
+			return false;
+		}
+		$app = $this->initializePradoApplication($args[1]);
+		TShellInterpreter::getInstance()->clearActionClass();
+		TShellInterpreter::getInstance()->addActionClass('Prado\\Shell\\Actions\\TAppHelpAction');
+		foreach ($app->getShellActionClasses() as $actions) {
+			TShellInterpreter::getInstance()->addActionClass($actions);
```

Comment:
This class basically resets the shell intepreter, loads a fixed set of actions and re-runs the shell interpreted.
I guess this would be somewhat used by an app to add its own actions to the shell interpreter, but how to do it?

Reply:
That was exactly the issue I was running into.  I decided to not worry about it *yet* and just do the refactor to see what ideas precipitated from it.  the Prado-cli would need to be redesigned.

eg. the first parameter would need to be the directory, or better yet have a -d or --directory input flag, otherwise the cli checks the current directory or 'protected' for the application.

I didn't know what you thought about re-doing the Shell and what metrics should be used.